src/hotspot/share/opto/divnode.cpp

```diff
@@ -504,6 +504,92 @@ Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
   return nullptr;
 }
 
+template<typename IntegerType>
+static const IntegerType* compute_generic_div_type(const IntegerType* i1, const IntegerType* i2, int widen) {
```

Comment:
Do we need the `generic` in the name? The `template` already suggests that it can be used for different types, right?

Also: I'm wondering if we can somehow extend this for `UDivI` and `UDIvL`.
I suppose you would have to use the `_ulo` and `_uhi` instead of `_lo` and `_hi`.

I'm not saying this all has to be done in this PR, but we could at least anticipate the extension to unsigned division.

Reply:
I've adjusted it to `compute_signed_div_type`.
I would not worry about unsigned division now and would leave this for a later RFE. Would you perhaps create a issue for that? I have no permission for the JBS